src/io/xpi.js

```diff
@@ -43,53 +43,51 @@ export class Xpi extends IOBase {
     }
     if (this.entries.includes(entry.fileName)) {
       log.info('Found duplicate file entry: "%s" in package', entry.fileName);
-      reject(new Error(oneLine`DuplicateZipEntry: Entry
-        "${entry.fileName}" has already been seen`));
+      throw new Error(oneLine`DuplicateZipEntry: Entry
+        "${entry.fileName}" has already been seen`);
     }
     this.entries.push(entry.fileName);
     this.files[entry.fileName] = entry;
   }
 
-  getFiles(_onEventsSubscribed) {
-    return new Promise((resolve, reject) => {
-      // If we have already processed the file and have data
-      // on this instance return that.
-      if (Object.keys(this.files).length) {
-        const wantedFiles = {};
-        Object.keys(this.files).forEach((fileName) => {
-          if (this.shouldScanFile(fileName)) {
-            wantedFiles[fileName] = this.files[fileName];
-          } else {
-            log.debug(`Skipping cached file: ${fileName}`);
-          }
-        });
-        return resolve(wantedFiles);
-      }
+  async getFiles(_onEventsSubscribed) {
+    // If we have already processed the file and have data
+    // on this instance return that.
+    if (Object.keys(this.files).length) {
+      const wantedFiles = {};
+      Object.keys(this.files).forEach((fileName) => {
+        if (this.shouldScanFile(fileName)) {
+          wantedFiles[fileName] = this.files[fileName];
+        } else {
+          log.debug(`Skipping cached file: ${fileName}`);
+        }
+      });
+      return wantedFiles;
+    }
 
-      return this.open()
-        .then((zipfile) => {
-          zipfile.on('entry', (entry) => {
-            this.handleEntry(entry, reject);
-          });
-
-          // When the last entry has been processed
-          // and the fd is closed resolve the promise.
-          // Note: we cannot use 'end' here as 'end' is fired
-          // after the last entry event is emitted and streams
-          // may still be being read with openReadStream.
-          zipfile.on('close', () => {
-            resolve(this.files);
-          });
-
-          if (_onEventsSubscribed) {
-            // Run optional callback when we know the event handlers
-            // have been inited. Useful for testing.
-            if (typeof _onEventsSubscribed === 'function') {
-              _onEventsSubscribed();
-            }
-          }
-        })
-        .catch(reject);
+    const zipfile = await this.open();
+
+    return new Promise((resolve) => {
+      zipfile.on('entry', (entry) => {
+        this.handleEntry(entry);
```

Comment:
`this.handleEntry(entry, reject);` here as previously described (same for "src/io/crx.js").

Reply:
I'm also wondering if we should also listen to zipfile.on('error', ...)` and call reject also in this case.

Reply:
@rpl do I need to add zipfile.on('error', ...)` and tests for this in this PR or we should open new issue for this?